File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/HookCommandRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires hook files to declare and run harness validation commands.
 */
public enum HookCommandRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "hookCommand";
    }
    private static final String CATEGORY = "hookCommand";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode allowedCmds = catNode.get("parameters").get("allowedCommands").get("maven");
        final String expectedCmd = allowedCmds == null || allowedCmds.size() == 0 ? "" : allowedCmds.get(0).asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        final String prePushPath = catNode.get("parameters").get("prePushHook").asText();
        final Path prePush = root.resolve(prePushPath);
        final List<Finding> prePushIssues = Stream.of(prePush)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .flatMap(p -> {
                    try {
                        final String prePushText = HarnessCheckHelper.readFile(root, p);
                        final String declaredCmd = extractHookCommand(prePushText);
                        return Stream.of(declaredCmd.isEmpty()
                                ? new Finding(severity, CATEGORY, "pre-push hook must declare Harness validation command")
                                : !expectedCmd.equals(declaredCmd)
                                ? new Finding(severity, CATEGORY, "pre-push hook declares unsupported validation command: " + declaredCmd)
                                : !prePushText.contains(declaredCmd)
                                ? new Finding(severity, CATEGORY, "pre-push hook must run the declared validation command")
                                : null);
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .filter(java.util.Objects::nonNull)
                .toList();
        final String preCommitPath = catNode.get("parameters").get("preCommitHook").asText();
        final Path preCommit = root.resolve(preCommitPath);
        final List<Finding> preCommitIssues = validatePreCommitHook(root, severity, preCommit);
        return Stream.concat(prePushIssues.stream(), preCommitIssues.stream()).toList();
    }

    private String extractHookCommand(String text) {
        return Stream.of(text.split("\\R"))
                .filter(line -> line.startsWith("# Harness validation command: "))
                .map(line -> line.substring("# Harness validation command: ".length()).trim())
                .findFirst()
                .orElse("");
    }

    private List<Finding> validatePreCommitHook(Path root, String severity, Path preCommit) throws MojoExecutionException {
        return Stream.of(preCommit)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .flatMap(p -> {
                    try {
                        final String preCommitText = HarnessCheckHelper.readFile(root, p);
                        final Pattern fullStackPattern = Pattern.compile("(^|\\s)(uv|bun|gradle|mvn)(\\s|$)|\\.\\./gradlew|harnessValidate|harness_validate\\.py|harness-validate\\.ts");
                        return fullStackPattern.matcher(preCommitText).find()
                                ? Stream.of(new Finding(severity, CATEGORY, "pre-commit hook must not run full stack validation commands"))
                                : Stream.empty();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/text/HookCommandRule.kt
package com.ririnto.sinon.harness.rules.text

import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.core.Severity
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText

/**
 * Rule that requires hooks to declare and run validation commands.
 *
 * Operates on plain text; the check uses literal substring matching for comment lines and requires no AST parser.
 */
object HookCommandRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "hookCommand"

    override fun validate(ctx: RuleContext): Collection<Finding> {
        val parametersObj =
            ctx.manifest
                .categoryObject(category)
                ?.get("parameters")
                ?.jsonObject ?: return emptyList()
        val severity = ctx.manifest.severityOf(category)
        val allowedPreCommitCmds =
            parametersObj["allowedPreCommitCommands"]?.jsonArray
                ?.mapNotNull { elem -> elem.jsonPrimitive.contentOrNull }
                ?: emptyList()
        val prePushHook = ctx.root / JsonAccess.stringFromObject(parametersObj, "prePushHook")
        val preCommitHook = ctx.root / JsonAccess.stringFromObject(parametersObj, "preCommitHook")
        return buildList {
            if (prePushHook.isRegularFile()) {
                addAll(
                    validateHook(
                        ctx,
                        prePushHook.readText(),
                        parametersObj["allowedCommands"]?.jsonArray
                            ?.mapNotNull { elem -> elem.jsonPrimitive.contentOrNull }
                            ?: emptyList(),
                        severity,
                        "pre-push hook",
                    ),
                )
            }
            if (preCommitHook.isRegularFile()) {
                addAll(
                    validateHook(
                        ctx,
                        preCommitHook.readText(),
                        allowedPreCommitCmds,
                        severity,
                        "pre-commit hook",
                        allowedPreCommitCmds.isNotEmpty(),
                    ),
                )
            }
        }
    }

    private fun validateHook(
        ctx: RuleContext,
        text: String,
        allowedCmds: List<String>,
        severity: Severity,
        hookName: String,
        requireDeclaration: Boolean = true,
    ): Collection<Finding> =
        buildList {
            val command = extractCommand(text)
            val missingMsg = { msgKey: String ->
                ctx.manifest.stringValue(category, msgKey).takeIf { msg -> msg.isNotEmpty() }
            }
            if (command.isEmpty() && requireDeclaration) {
                add(
                    Finding(
                        severity,
                        category,
                        missingMsg("missingDeclaration") ?: "$hookName must declare Harness validation command",
                    ),
                )
            }
            if (command.isNotEmpty() && command !in allowedCmds) {
                add(
                    Finding(
                        severity,
                        category,
                        missingMsg("unsupportedCommand")
                            ?: "$hookName declares unsupported validation command: $command",
                    ),
                )
            }
            if (command.isNotEmpty() && !text.contains(command)) {
                add(
                    Finding(
                        severity,
                        category,
                        missingMsg("commandNotRun") ?: "$hookName must run the declared validation command",
                    ),
                )
            }
        }

    private fun extractCommand(text: String): String =
        text
            .lineSequence()
            .firstOrNull { line -> line.startsWith("# Harness validation command: ") }
            ?.removePrefix("# Harness validation command: ")
            ?.trim()
            ?: ""
}
